### PR TITLE
Implement thread pool and thread safe queue

### PR DESCRIPTION
**Pull request by NazariiProshyn**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 3.16.3)
+project(ThreadPoolC17)
+
+set(CMAKE_CXX_STANDART 17)
+
+add_executable(ThreadPoolC17 Main.cpp ThreadPoolH.h ThreadPoolH.cpp  ThreadQueue.h ThreadQueue.cpp)
```
Comment (mlohvynenko-clgx) [inline CMakeLists.txt]: add an empty line in each file

File: ThreadPoolH.cpp [hunk 2]
```diff
@@ -0,0 +1,10 @@
+#include "ThreadPoolH.h"
+
+
+inline ThreadPool::ThreadPool(size_t NumOfThreads)
```
Comment (mlohvynenko-clgx) [inline ThreadPoolH.cpp]: read more about inline

File: CMakeLists.txt (hunk 1 above)
Comment (mlohvynenko-clgx) [reply CMakeLists.txt]: file naming convention:
use such style main.cpp, threadPool

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+cmake_minimum_required(VERSION 3.16.3)
+project(ThreadPoolC17)
```
Comment (mlohvynenko) [inline CMakeLists.txt:2]: lower case.
threadPool

File: ThreadPoolH.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include <vector>
+#include <queue>
+#include <functional> //bind
+#include <mutex>
+#include <condition_variable>
+#include <type_traits> //invoke_result
+#include <thread>
+#include <future> //packaged_task
```
Comment (mlohvynenko-clgx) [inline ThreadPoolH.h]: do not comment includes

File: ThreadPoolH.h [hunk 5]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+
```
Comment (mlohvynenko-clgx) [inline ThreadPoolH.h]: add doxxxygen file description

Comment (mlohvynenko-clgx) [reply ThreadPoolH.h]:
```
///
/// \file            <filename>
/// \date            <date>
/// \author          <your name>
/// \brief            <short description>
///
```

File: ThreadQueue.h [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+#include <thread>
+#include <condition_variable>
+#include <queue>
+#include <utility>
+
+template<typename T>
+class ThreadQueue
+{
+public:
+	ThreadQueue(T value);
```
Comment (mlohvynenko-clgx) [inline ThreadQueue.h]: what is the benefit of this constructor?

File: ThreadQueue.cpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+
+bool ThreadQueue::empty() const
+{
+	std::lock_guard<std::mutex> lock(mut);
+	return dataQueue.empty();
+}
+
+/*
```
Comment (nproshyn-CLGX) [inline ThreadQueue.cpp]: delete

File: joinThread.h [hunk 8]
```diff
@@ -0,0 +1,31 @@
+//	$Header$
```
Comment (mlohvynenko-clgx) [inline joinThread.h]: Header files should start with pragma once or ifndef 
Make a rule of it

File: threadQueue.h [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+	ThreadQueue() = default;
+
+	//! \brief                 Inserts a new element at the end of the queue
+	//! \param [in] value      New item for push 
+	//! \return 
+	void push(std::function<void()> value);
+
+	//! \brief                 Removes the next element in the queue after    
+	//!                        checking for items
+	//! \return                Shares pointer to the beginning of the queue
```
Comment (mlohvynenko-clgx) [inline threadQueue.h]: Is there a typo?
shared pointer? not shares

File: threadQueue.h [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+	std::shared_ptr<std::function<void()>> Pop();
+
+	//! \brief                     Assigns first element of queue to the variable
+	//! \param [in,out] value      function performed in the thread 
+	//! \return                    returns result of deleting first element
+	bool Pop(std::function<void()>& value);
```
Comment (mlohvynenko-clgx) [inline threadQueue.h]: Why don't you mention that pop return the first element from the queue, and it also removes the first element from the queue
And you mentioned nothing about blocking (mutext)

File: threadQueue.h [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+private:
+	//Allows you to work with the 
+	//queue in multithreaded mode safely
+	mutable std::mutex mut;
+
+	//Contains items about functions that 
+	//are in runtime and that will be executed
+	std::queue<std::function<void()>> dataQueue;
+
+	std::condition_variable dataCondition;
```
Comment (mlohvynenko-clgx) [inline threadQueue.h]: Why do you need this condition variable.

File: threadQueue.h [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+//                      the ability to simultaneously write
+//                      from multiple threads
+//	********************************************************
+
+
+#pragma once
+
+#include <mutex>
+#include <memory>
+#include <thread>
```
Comment (mlohvynenko-clgx) [inline threadQueue.h]: why did you include thread, utility?

File: joinThread.h [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+#include <thread>
+
+class JoinThread
+{
+public:
+	explicit JoinThread(std::vector<std::thread>& thread) :
+		threads(thread)
+	{};
+
+	//! \brief                Ensures the connection of all threads
```
Comment (mlohvynenko-clgx) [inline joinThread.h]: Bad comment. I do not understand it

File: threadPool.cpp [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+
+
+ThreadPool::ThreadPool():
+	done(false), joiner(threads)
+{
+	unsigned const thread_count =
+		std::thread::hardware_concurrency();
+	try 
+	{
+		for (unsigned i = 0; i < thread_count; i++)
```
Comment (mlohvynenko-clgx) [inline threadPool.cpp]: prefer size_t

File: threadPool.cpp [hunk 15]
```diff
@@ -0,0 +1,48 @@
+#include "threadPool.h"
+
+
+ThreadPool::ThreadPool():
+	done(false), joiner(threads)
+{
+	unsigned const thread_count =
```
Comment (mlohvynenko-clgx) [inline threadPool.cpp]: const size_t  hardwareConc = std::thread::hardware_concurrency();
threads.reserve(hardwareConc); // read more about reserve

File: threadPool.h [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+	std::vector<std::thread> threads;
+
+	//Indicates the possibility threads to work
+	std::atomic_bool done;
+
+	//Queue of functions to be performed
+	ThreadQueue workQueue;
+
+	//Ensures the connection of all threads
+	JoinThread joiner;
```
Comment (mlohvynenko-clgx) [inline threadPool.h]: Why do not you just join all your threads in the destructor by yourselfe?

File: threadQueue.cpp [hunk 17]
```diff
@@ -0,0 +1,42 @@
+#include "threadQueue.h"
+
+
+void ThreadQueue::push(std::function<void()> value)
+{
+	std::lock_guard<std::mutex> lock(mut);
+	dataQueue.push(std::move(value));
+	dataCondition.notify_one();
```
Comment (mlohvynenko-clgx) [inline threadQueue.cpp]: And who is waiting for this notification?

File: threadPool.cpp [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+	{
+		for (unsigned i = 0; i < thread_count; i++)
+		{
+			threads.push_back(
+				std::thread(&ThreadPool::workThread, this));
+		}
+	}
+	catch (...)
+	{
+		done = true;
```
Comment (mlohvynenko) [inline threadPool.cpp:19]: What is the purpose of dropping this flag to true\false when the constructor raises an exception?

File: main.cpp [hunk 19]
```diff
@@ -0,0 +1,32 @@
+#include "threadPool.h"
+#include "windows.h"
```
Comment (mlohvynenko-clgx) [inline main.cpp]: Do not use platform-specific headers.

File: main.cpp [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+{
+    ThreadPool threadpool(4);
+    std::function<void()> f1 = foo1;
+    std::function<void()> f2 = foo2;
+    std::function<void()> f3 = foo3;
+
+    for (size_t i = 0; i < 3; ++i) 
+    { 
+        threadpool.submit(f1);
+        threadpool.submit(f2);
```
Comment (mlohvynenko-clgx) [inline main.cpp]: why not threadpool.submit(foo1) ?

File: main.cpp [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+        threadpool.submit(f3);
+        Sleep(2000);
```
Comment (mlohvynenko-clgx) [inline main.cpp]: bad practice. Do not use platform-specific stuff if a standardized alternative exists.

File: threadPool.h [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+
+#include "threadQueue.h"
+
+class ThreadPool
+{
+public:
+	//! \brief       Determines the possible number 
+	//!              of threads and creates them
+	//! \return 
+	ThreadPool(size_t thread_count);
```
Comment (mlohvynenko-clgx) [inline threadPool.h]: const size_t

File: threadPool.h [hunk 23]
```diff
@@ -0,0 +1,56 @@
...
+
+	~ThreadPool();
+
+	//Thread not copiable
+	ThreadPool(const ThreadPool&) = delete;
+	ThreadPool& operator = (const ThreadPool&) = delete;
```
Comment (mlohvynenko-clgx) [inline threadPool.h]: Remove space `operator= `

File: threadPool.h [hunk 24]
```diff
@@ -0,0 +1,56 @@
...
+	//! \param [in] funct      Function performed in the thread 
+	//! \return                   
+	void submit(std::function<void()> funct);
+
+private:
+	//Vector of working threads
+	std::vector<std::thread> threads;
+
+	//Indicates the possibility threads to work
+	std::atomic_bool run = false;
```
Comment (mlohvynenko-clgx) [inline threadPool.h]: std::atomic_bool run{false};

File: threadPool.cpp [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+	{
+		run = true;
+		std::cout << e.what() << std::endl;
+		throw;
+	}
+}
+
+ThreadPool::~ThreadPool()
+{
+	//while (workQueue.size() > 0) {}
```
Comment (mlohvynenko-clgx) [inline threadPool.cpp]: remove commented code from your PR

File: threadPool.cpp [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+	run = true;
+	for (size_t i = 0; i < threads.size(); ++i)
```
Comment (mlohvynenko-clgx) [inline threadPool.cpp]: Use iterators.

File: threadQueue.h [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+	//! \brief                     Assigns first element of queue to the variable
+	//!                            blocks access to the queue
+	//! \param [in,out] value      function performed in the thread 
+	//! \return                    returns result of deleting first element
+	std::function<void()> pop();
+
+	//! \brief                 Check queue for items     
+	//!                        blocks access to the queue
+	//! \return                Result of checking (true if number of items>0)
+	bool empty() const;
```
Comment (mlohvynenko-clgx) [inline threadQueue.h]: why do you need this method?
It's useless.

File: threadQueue.cpp [hunk 28]
```diff
@@ -0,0 +1,22 @@
+#include "threadQueue.h"
+
+
+void ThreadQueue::push(const std::function<void()>& value)
```
Comment (mlohvynenko-clgx) [inline threadQueue.cpp]: const ref doesn't work with move semantic/perfect forwarding

File: threadQueue.h [hunk 29]
```diff
@@ -0,0 +1,51 @@
...
+	//! \return 
+	void push(const std::function<void()>& value);
+
+	//! \brief                     Assigns first element of queue to the variable
+	//!                            blocks access to the queue
+	//! \param [in,out] value      function performed in the thread 
+	//! \return                    returns result of deleting first element
+	std::function<void()> pop();
+
+	size_t size() { return dataQueue.size(); }
```
Comment (mlohvynenko-clgx) [inline threadQueue.h]: size_t size() const !

File: threadPool.cpp [hunk 30]
```diff
@@ -0,0 +1,48 @@
...
+	threads.reserve(thread_count);
+	try
+	{
+		for (size_t i = 0; i < thread_count; i++)
+		{
+			threads.push_back(
+				std::thread(&ThreadPool::workThread, this));
+		}
+	}
+	catch (std::exception& e)
```
Comment (mlohvynenko-clgx) [inline threadPool.cpp]: const

**Pull request merged**